samples: net: add secure MQTT sensor/actuator device sample

jpmur opened the pull request:
This sample demonstrates the implementation of an (industrial) IoT sensor/actuator device. The application uses the MQTT protocol to securely send sensor data to a remote MQTT broker, while responding to commands received over the MQTT connection.

The application serves as a demonstration of the integration of a few Zephyr subsystems, namely the networking and sensor subsystems in a system-level solution.

Tested with `adi_eval_adin1110ebz` (using Ethernet as the network L1/L2).

See README.rst for full details.

cc @MaureenHelm

jpmur committed:
samples: net: add secure MQTT sensor/actuator device sample

This sample demonstrates the implementation of an (industrial) IoT
sensor/actuator device. The application uses the MQTT protocol to
securely send sensor data to a remote MQTT broker, while responding
to commands received over the MQTT connection.

Signed-off-by: Jason Murphy <jason.murphy@analog.com>

jukkar commented on samples/net/secure_mqtt_sensor_actuator/overlay_static_insecure.conf:
```diff
@@ -0,0 +1,18 @@
+
+# Static IP address config
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="169.254.0.2"
+CONFIG_MQTT_BROKER_IPV4_ADDR="169.254.0.1"
```
Please use the IP address space we use in other samples (the addresses are reserved for documentation use).
So here
```
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
CONFIG_NET_SAMPLE_MQTT_BROKER_IPV4_ADDR="192.0.2.2"
```

jpmur replied:
Done.

jukkar commented on samples/net/secure_mqtt_sensor_actuator/src/main.c:
```diff
@@ -0,0 +1,86 @@
+	mac = net_if_get_link_addr(iface);
+
+	LOG_INF("MAC Address: %02X:%02X:%02X:%02X:%02X:%02X",
+			mac->addr[0], mac->addr[1], mac->addr[3],
+			mac->addr[3], mac->addr[4], mac->addr[5]);
```
Please fix indentation

jpmur replied:
Done.

jukkar reviewed (commented):
Some initial comments at this point.

Side note to @rlubos. After this one is merged, there will be 4 mqtt related samples so I think we should consider placing various MQTT samples under `samples/net/mqtt` directory.

MaureenHelm commented on samples/net/secure_mqtt_sensor_actuator/boards/adi_eval_adin1110ebz.overlay:
```diff
@@ -0,0 +1,26 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+	aliases {
+		ambient-temp0 = &adt7420;
+	};
+};
+
+&i2c3 {
+	/* Add node label to ADT7420 */
+	adt7420: adt7420@48 {};
+};
```
These can be removed now that #70707 has been merged

jpmur replied:
Done.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/README.rst:
```diff
@@ -0,0 +1,185 @@
+.. zephyr-app-commands::
+	 :zephyr-app: samples/net/secure_mqtt_sensor_actuator
+	 :board: adi_eval_adin1110ebz
+	 :conf: overlay_static_insecure.conf
```
This command is wrong, provides an overlay file as a config file (I. e. `-DCONF_FILE` instead of `-DOVELRAY_CONFIG`).

jpmur replied:
I think I have fixed this - can you check if it is correct now?

rlubos replied:
Looks ok now

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/tls_config/cert.h:
```diff
@@ -0,0 +1,45 @@
+ */
+
+/* CA certificate for test.mosquitto.org
+ * (valid at time of development, may require updating)
```
Would be nice to include source of this certificate.

jpmur replied:
I added the source - test.mosquitto.org

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/tls_config/config-tls-app.h:
```diff
@@ -0,0 +1,40 @@
+ * This file is included at the end of Zephyr's generic configuration.
+ */
+
+#define MBEDTLS_PK_HAVE_ECC_KEYS
```
Are all those configs really needed? First, we should aim to configure mbed TLS with Kconfig only (pretty much of those can be configured in prj.conf), but what's more important, the `mqtt_publisher` sample seems to be able to establish TLS connection with `test.mosquitto.org` running pretty much on defaults:
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/mqtt_publisher/overlay-tls.conf#L5
The mosquito certificate seems to rely on RSA encryption (enabled by default) not EC.

jpmur replied:
You're right - not all needed. These were the configs I needed for the particular certs I was using with a local Mosquitto instance. I opted to leave them there to serve as a reference as to how you can configure MbedTLS in your app.
I realise is is probably best to disable any unneeded features so I removed this extra config file and Kconfig symbols.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/mqtt_client.c:
```diff
@@ -0,0 +1,562 @@
+void mqtt_client_thread(void)
+{
+	/* Wait for network to come up */
+	while (k_sem_take(&net_conn_sem, K_MSEC(MSECS_NET_POLL_TIMEOUT)) != 0) {
```
IMO this could pretty much continue on the main thread, no need to start a separate thread.

jpmur replied:
Good point - was an early design decision that I never looked back on.
I moved the MQTT client functionlity to the main thread.
I created a new header file to expose the MQTT client APIs in mqtt_client.c
Increased main thread stack size from 1024 to 2048.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/mqtt_client.c:
```diff
@@ -0,0 +1,562 @@
+	tls_config->cipher_list = NULL;
+	tls_config->sec_tag_list = m_sec_tags;
+	tls_config->sec_tag_count = ARRAY_SIZE(m_sec_tags);
+	#if defined(CONFIG_MBEDTLS_SERVER_NAME_INDICATION)
```
Preprocessor directives shouldn't be indented.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/mqtt_client.c:
```diff
@@ -0,0 +1,562 @@
+	}
+}
+
+K_TIMER_DEFINE(publish_timer, publish_timer_handler, NULL);
```
Just use `k_work_delayable` instead of using work and a timer separately.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/mqtt_client.c:
```diff
@@ -0,0 +1,562 @@
+	int rc;
+
+	rc = poll_mqtt_socket(client, mqtt_keepalive_time_left(client));
+
```
No need for an empty line here.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/mqtt_client.c:
```diff
@@ -0,0 +1,562 @@
+int app_mqtt_subscribe(struct mqtt_client *client)
+{
+	int rc;
+
```
No need for empty lines around variable definitions.

jpmur replied:
Done.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/device.h:
```diff
@@ -0,0 +1,56 @@
+/** @brief Device command consisting of a command string
+ *         and associated handler function pointer
+ */
+typedef struct device_cmd {
```
Please don't use typedefs, it's against Zephyr coding guidelines.

jpmur replied:
Done.

rlubos reviewed (commented):
I have one major issue with this sample - in current form it's quite HW dependent for a generic networking sample. Moreover, it's not even possible to build for simulated platforms (qemu_x86 or native_sim) which makes the future maintenance of the sample, complicated.

rlubos commented:
> After this one is merged, there will be 4 mqtt related samples so I think we should consider placing various MQTT samples under samples/net/mqtt directory.

@jukkar I think we could do it, however so far I've counted only two (not counting the one from this PR) - one for standard MQTT and one for MQTT-SN? Am I missing something?

jukkar commented:
> I think we could do it, however so far I've counted only two (not counting the one from this PR) - one for standard MQTT and one for MQTT-SN? Am I missing something?

Indeed, my bad. There was two of them, sorry about that.

jpmur commented:
> > I think we could do it, however so far I've counted only two (not counting the one from this PR) - one for standard MQTT and one for MQTT-SN? Am I missing something?
> 
> Indeed, my bad. There was two of them, sorry about that.

`samples/net/cloud/aws_iot_mqtt` and `samples/net/cloud/mqtt_azure` may have been the ones you were thinking of?

jpmur commented:
> I have one major issue with this sample - in current form it's quite HW dependent for a generic networking sample. Moreover, it's not even possible to build for simulated platforms (qemu_x86 or native_sim) which makes the future maintenance of the sample, complicated.

I understand that concern. The intention of the sample is to be not a purely networking focused sample, but a demonstration of networking + sensor etc. Obviously, it is primarily network focused so having it in `samples/net` seemed like the best option.

I think it is pretty easy to get this sample running on any network-capable board using devicetree to add the sensor alias. Most boards have LEDs too.

Perhaps it belongs in a different location?

kartben commented:
> Perhaps it belongs in a different location?

Related to #67021. It feels like this "sample" could be a great candidate for a first new "demo" app.

rlubos commented:
> I understand that concern. The intention of the sample is to be not a purely networking focused sample, but a demonstration of networking + sensor etc. Obviously, it is primarily network focused so having it in `samples/net` seemed like the best option.
> 
> I think it is pretty easy to get this sample running on any network-capable board using devicetree to add the sensor alias. Most boards have LEDs too.

I was wondering if it'd be possible to use `DEVICE_DT_GET_OR_NULL()` for instance, so that the we can compile the sample even if led/sensor are missing, and then "fake" the behavior if we don't get the device pointer. Provide dummy sensor value or print log in case led event is triggered. That way we could still run the sample in simulated environments, which is beneficial when investigating purely networking issues.

jpmur commented:
> > Perhaps it belongs in a different location?
> 
> Related to #67021. It feels like this "sample" could be a great candidate for a first new "demo" app.

Yeah, I think this sample is suited to that type of application.
Is there any plan to follow up on what was discussed in that RFC?

jpmur commented on samples/net/secure_mqtt_sensor_actuator/Kconfig:
```diff
@@ -0,0 +1,62 @@
+
+config MQTT_BROKER_IPV4_ADDR
+	string "IPv4 Address of MQTT Broker"
+	default "91.121.93.94"
```
Done. 
I opted to use the getaddrinfio() Posix API.
I opted to remove `CONFIG_MQTT_BROKER_IPV4_ADDR` in favour of just a Kconfig symbol for the MQTT broker's hostname. User's can set this symbol to be the hostname or IP address, and works fine.

jpmur commented on samples/net/secure_mqtt_sensor_actuator/Kconfig:
```diff
@@ -0,0 +1,62 @@
+
+mainmenu "Secure MQTT Sensor Actuator Sample Application"
+
+config MQTT_BROKER_IPV4_ADDR
```
Done.

jpmur commented on samples/net/secure_mqtt_sensor_actuator/Kconfig:
```diff
@@ -0,0 +1,62 @@
+	help
+	  MQTT_QOS_0_AT_MOST_ONCE = 0
+	  MQTT_QOS_1_AT_LEAST_ONCE = 1
+	  MQTT_QOS_2_EXACTLY_ONCE = 2
```
The benefit of the current implementation is that the Kconfig symbol for QoS is an int and its value translates directly over to the enum type for QoS in the MQTT lib. This allows the symbol to be used directly in the code to set the MQTT QoS, which makes things nice and simple.

Is using `choice` a must-have?

jukkar replied:
It would be better if we do not force the user to know what each value means.
You can build you code such a way that the mapping from this config symbol to int value is done at built time.
```
IS_ENABLED(CONFIG_NET_SAMPLE_MQTT_QOS_AT_MOST_ONCE) ? 0 : 
(IS_ENABLED(CONFIG_NET_SAMPLE_MQTT_QOS_AT_LEAST_ONCE) ? 1 : 2)
```

jpmur commented on samples/net/secure_mqtt_sensor_actuator/Kconfig:
```diff
@@ -0,0 +1,62 @@
+	  MQTT_QOS_1_AT_LEAST_ONCE = 1
+	  MQTT_QOS_2_EXACTLY_ONCE = 2
+
+config MQTT_SUBSCRIBE_QOS
```
As above.

jpmur commented on samples/net/secure_mqtt_sensor_actuator/README.rst:
```diff
@@ -0,0 +1,185 @@
+Overview
+********
+Secure MQTT Sensor/Actuator Sample Application
```
Done.

jpmur commented on samples/net/secure_mqtt_sensor_actuator/README.rst:
```diff
@@ -0,0 +1,185 @@
+Secure MQTT Sensor/Actuator Sample Application
+
+This sample demonstrates the implementation of an IoT sensor/actuator device.
+The application uses the MQTT protocol to securely send sensor data to a remote MQTT broker, while responding to commands
```
Do you mean to split the sentence up with punctuation or just have less words on each line?

jukkar replied:
> just have less words on each line?

This one

jpmur commented on samples/net/secure_mqtt_sensor_actuator/prj.conf:
```diff
@@ -0,0 +1,56 @@
+CONFIG_NET_CONFIG_NEED_IPV4=y
+
+# Disable IPv6
+CONFIG_NET_IPV6=n
```
No particular reason, I just was more familiar with using IPv4 addresses.
I enabled IPv6.

jukkar replied:
In Zephyr we try to support both IPv4 and IPv6 in the code and samples, that is why I was asking that.

jpmur commented on samples/net/secure_mqtt_sensor_actuator/overlay_static_insecure.conf:
```diff
@@ -0,0 +1,18 @@
+# Static IP address config
+CONFIG_NET_CONFIG_SETTINGS=y
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="169.254.0.2"
+CONFIG_MQTT_BROKER_IPV4_ADDR="169.254.0.1"
```
Have removed this symbol anyway, as mentioned above.

jpmur commented on samples/net/secure_mqtt_sensor_actuator/overlay_static_insecure.conf:
```diff
@@ -0,0 +1,18 @@
+# Config to disable TLS and DHCPv4, allowing insecure MQTT and a static IP address.
+# This allows quick testing of the application without need for a DHCP server or secure MQTT broker set up.
+# Only recommended for quick sampling/testing.
+# Usage: west build -b <board> -- -DOVERLAY_CONFIG=overlay_static_insecure.conf
```
Done.

jpmur commented:
> > I understand that concern. The intention of the sample is to be not a purely networking focused sample, but a demonstration of networking + sensor etc. Obviously, it is primarily network focused so having it in `samples/net` seemed like the best option.
> > I think it is pretty easy to get this sample running on any network-capable board using devicetree to add the sensor alias. Most boards have LEDs too.
> 
> I was wondering if it'd be possible to use `DEVICE_DT_GET_OR_NULL()` for instance, so that the we can compile the sample even if led/sensor are missing, and then "fake" the behavior if we don't get the device pointer. Provide dummy sensor value or print log in case led event is triggered. That way we could still run the sample in simulated environments, which is beneficial when investigating purely networking issues.

Yeah I should be able to do that. Something similar is done in the `lwm2m_client` sample.
That sample uses a temperature sensor also.

jpmur commented:
@rlubos @jukkar 
Apologies for the delay - was out of office for a while.
I made the last of the requested changes re. spoofing the sensor value.
All other comments have been addressed too.

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/device.c:
```diff
@@ -0,0 +1,128 @@
+
+/* Devices */
+static const struct device *sensor = DEVICE_DT_GET_OR_NULL(DT_ALIAS(ambient_temp0));
+static const struct device *leds = DEVICE_DT_GET_ONE(gpio_leds);
```
Would need to use `DEVICE_DT_GET_OR_NULL()` for leds as well, currently it won't build for emulated platforms.

jpmur replied:
Done

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/device.c:
```diff
@@ -0,0 +1,128 @@
+		}
+	}
+
+	if (!device_is_ready(leds)) {
```
And check for `leds != NULL` here.

jpmur replied:
Done

rlubos commented on samples/net/secure_mqtt_sensor_actuator/src/device.c:
```diff
@@ -0,0 +1,128 @@
+{
+	int rc;
+
+	switch (state) {
```
And here, for example something like this could fake LED behavior if the actual leds are not available:
```c
        switch (state) {
        case LED_OFF:
+               if (leds == NULL) {
+                       LOG_INF("LED %d OFF", led_idx);
+                       break;
+               }
+
                led_off(leds, led_idx);
                break;
        case LED_ON:
+               if (leds == NULL) {
+                       LOG_INF("LED %d ON", led_idx);
+                       break;
+               }
+
                led_on(leds, led_idx);
                break;
```

jpmur replied:
Done

rlubos commented on samples/net/secure_mqtt_sensor_actuator/overlay-static-insecure.conf:
```diff
@@ -0,0 +1,18 @@
+# Config to disable TLS and DHCPv4, allowing insecure MQTT and a static IP address.
```
Could add a simple `overlay-static.conf` as well, so that we can test the sample in more or less default configuration (co communicating securely with public broker), but with static IP addresses. The following config worked out-of-the-box for native_sim or qemu_x86:
```
CONFIG_NET_DHCPV4=n
CONFIG_NET_CONFIG_SETTINGS=y
CONFIG_NET_CONFIG_NEED_IPV4=y
CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
CONFIG_NET_CONFIG_MY_IPV4_GW="192.0.2.2"
CONFIG_DNS_RESOLVER=y
CONFIG_DNS_SERVER_IP_ADDRESSES=y
CONFIG_DNS_SERVER1="192.0.2.2"
```

jpmur replied:
I tested this on hardware, connecting to the public Mosquitto broker and the connection timed out.

rlubos replied:
This is likely because of gateway/dns server addresses, which need to be set accordingly to your network configuration. This example was tailored for the networks we use with native_sim/qemu.

rlubos replied:
DNS server could likely be set to some public one (like `8.8.8.8`), but w/o setting correct gateway address for your local network, you won't be able to reach the internet.

jpmur replied:
Thanks @rlubos 👍

rlubos commented on samples/net/secure_mqtt_sensor_actuator/sample.yaml:
```diff
@@ -0,0 +1,22 @@
+    filter: dt_alias_exists("ambient-temp0")
+    integration_platforms:
+      - adi_eval_adin1110ebz
+  sample.net.secure_mqtt_sensor_actuator.staticip_insecure:
```
And finally, would be good to add one build configuration for `native_sim` for example.

jpmur replied:
Done

rlubos reviewed (commented):
Thanks, a few minor tweaks were needed to run the sample with `native_sim` or `qemu_x86`, see comments below, otherwise looks ok. Would be good to apply them to make the sample fully platform-independent.

jpmur commented:
@MaureenHelm could you have a look again please?

kartben commented on samples/net/secure_mqtt_sensor_actuator/README.rst:
```diff
@@ -0,0 +1,204 @@
+.. zephyr:code-sample:: secure-mqtt-sensor-actuator
+   :name: Secure MQTT Sensor Actuator
```
```suggestion
   :name: Secure MQTT Sensor/Actuator
```

kartben commented on samples/net/secure_mqtt_sensor_actuator/README.rst:
```diff
@@ -0,0 +1,204 @@
+	[00:00:12.769,000] <inf> app_mqtt: topic: 'zephyr_sample/command', payload: led_on
+	[00:00:12.770,000] <inf> app_device: Executing device command: led_on
+
+.. _Mosquitto: https://mosquitto.org/download/
```
```suggestion
.. _Eclipse Mosquitto: https://mosquitto.org/download/
```

kartben commented on samples/net/secure_mqtt_sensor_actuator/Kconfig:
```diff
@@ -0,0 +1,58 @@
+	  This config determines the frequency at which MQTT publishes occur.
+
+choice NET_SAMPLE_MQTT_QOS
+	prompt "Quality of Service used for MQTT publish and subscribe"
```
```suggestion
	prompt "Quality of Service level used for MQTT publish and subscribe"
```

kartben commented on samples/net/secure_mqtt_sensor_actuator/Kconfig:
```diff
@@ -0,0 +1,58 @@
+	prompt "Quality of Service used for MQTT publish and subscribe"
+	default NET_SAMPLE_MQTT_QOS_1_AT_LEAST_ONCE
+
+config NET_SAMPLE_MQTT_QOS_0_AT_MOST_ONCE
+	bool "Lowest Quality of Service"
+	help
+	  No acknowledgment needed for published message.
+
+config NET_SAMPLE_MQTT_QOS_1_AT_LEAST_ONCE
+	bool "Medium Quality of Service"
+	help
+	  If acknowledgment expected for published message, duplicate messages permitted.
+
+config NET_SAMPLE_MQTT_QOS_2_EXACTLY_ONCE
+	bool "Highest Quality of Service"
+	help
+	   Acknowledgment expected and message shall be published only once.
```
really not a fan of coming up with new terminology (lowest/medium/highest) for the different QoS levels and recommend to just use the descriptions from the standard

```suggestion
config NET_SAMPLE_MQTT_QOS_0_AT_MOST_ONCE
	bool "QoS 0 / At most once delivery"
	help
	  No acknowledgment needed for published message.

config NET_SAMPLE_MQTT_QOS_1_AT_LEAST_ONCE
	bool "QoS 1 / At least once delivery"
	help
	  If acknowledgment expected for published message, duplicate messages permitted.

config NET_SAMPLE_MQTT_QOS_2_EXACTLY_ONCE
	bool "QoS 2 / Exactly once delivery"
	help
	   Acknowledgment expected and message shall be published only once.
```

kartben reviewed (changes requested):
Did a quick pass on the documentation. 
This sample "on steroids"  will likely be a good candidate for when we will have found a good home for https://github.com/zephyrproject-rtos/zephyr/issues/67021, but it's certainly OK with me to have it live alongside network samples for now -- thanks for your work on this!

kartben reviewed (changes requested):
Did a quick pass on the documentation. 
This sample "on steroids"  will likely be a good candidate for when we will have found a good home for https://github.com/zephyrproject-rtos/zephyr/issues/67021, but it's certainly OK with me to have it live alongside network samples for now -- thanks for your work on this!

jpmur commented:
> Did a quick pass on the documentation. This sample "on steroids" will likely be a good candidate for when we will have found a good home for #67021, but it's certainly OK with me to have it live alongside network samples for now -- thanks for your work on this!

Thank you for the review! Have addressed your comments.

kartben commented on samples/net/secure_mqtt_sensor_actuator/README.rst:
```diff
@@ -0,0 +1,219 @@
+
+There are a few ways to configure the application:
+
+.. list-table::
+
+		* - :file:`prj.conf`
+			- Default config: Secure MQTT, dynamic IP address (DHCP)
+
+		* - :file:`overlay-static.conf`
+			- Secure MQTT, static IP address
+
+		* - :file:`overlay-static-insecure.conf`
+			- Unsecure MQTT, static IP address
```
this is not correctly formatted (see [output doc](http://127.0.0.1:8080/samples/net/secure_mqtt_sensor_actuator/README.html#software-setup)) , probably due to incorrect tabulations

```suggestion
.. list-table::

   * - :file:`prj.conf`
     - Default config: Secure MQTT, dynamic IP address (DHCP)

   * - :file:`overlay-static.conf`
     - Secure MQTT, static IP address

   * - :file:`overlay-static-insecure.conf`
     - Unsecure MQTT, static IP address
```

kartben reviewed (changes requested):
one last request re: docs as one of your latest changes doesn't render correctly - thanks again!

jpmur commented:
> one last request re: docs as one of your latest changes doesn't render correctly - thanks again!

@kartben Documentation build is failing now. Can you take a look? Not sure how to interpret the issue - looks like its failing to checkout the repo..

kartben commented:
> > one last request re: docs as one of your latest changes doesn't render correctly - thanks again!
> 
> @kartben Documentation build is failing now. Can you take a look? Not sure how to interpret the issue - looks like its failing to checkout the repo..

GitHub was acting up earlier today - I will try to re-trigger CI
